server/blindai_sgx/src/model_store.rs

```diff
@@ -0,0 +1,141 @@
+use anyhow::{anyhow, Result};
+use log::*;
+use ring::digest::{self, Digest};
+#[cfg(not(target_env = "sgx"))]
+use std::sync::RwLock;
+#[cfg(target_env = "sgx")]
+use std::sync::SgxRwLock as RwLock;
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    sync::Arc,
+};
+use uuid::Uuid;
+
+use crate::model::{InferenceModel, ModelDatumType, OnnxModel};
+
+struct InnerModelStore {
+    models_by_id: HashMap<Uuid, InferenceModel>,
+    onnx_by_hash: HashMap<Vec<u8>, (usize, Arc<OnnxModel>)>,
+}
+
+/// This is where model are stored.
+pub struct ModelStore {
+    inner: RwLock<InnerModelStore>,
+}
+
+impl ModelStore {
+    pub fn new() -> Self {
+        ModelStore {
+            inner: RwLock::new(InnerModelStore {
+                models_by_id: HashMap::new(),
+                onnx_by_hash: HashMap::new(),
+            }),
+        }
+    }
+
+    pub fn add_model(
+        &self,
+        model_bytes: &[u8],
+        input_fact: Vec<usize>,
+        datum: ModelDatumType,
+        model_name: Option<String>,
+        datum_output: ModelDatumType,
+    ) -> Result<(Uuid, Digest)> {
+        let model_id = Uuid::new_v4();
+        let model_hash = digest::digest(&digest::SHA256, &model_bytes);
+
+        let model_hash_vec = model_hash.as_ref().to_vec();
+
+        // Create an entry in the hashmap and in the dedup map
+        {
+            // take the write lock
+            let mut models = self.inner.write().unwrap();
+
+            // HashMap entry api requires only one lookup and should be prefered than .get()
+            // followed with .insert()
+
+            // deduplication support
+            let model = match models.onnx_by_hash.entry(model_hash_vec.clone()) {
+                Entry::Occupied(mut entry) => {
+                    let (num, onnx) = entry.get_mut();
+                    *num += 1;
+                    info!("Reusing an existing ONNX entry for model. (n = {})", *num);
```

Comment:
This feature might be useful for users, but beware if there are multiple model uploaders. You actually leak what models others have previously updated...